src/tokens/fungible.rs

```diff
@@ -0,0 +1,528 @@
+//! The fungible token.
+
+use super::types::{Amount, Operator, Owner};
+use gstd::{prelude::*, ActorId};
+use hashbrown::HashMap;
+
+#[cfg(test)]
+use super::test_helper::msg;
+#[cfg(not(test))]
+use gstd::msg;
+
+pub mod encodable;
+pub mod extensions;
+
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
+struct OwnerData {
+    balance: Amount,
+    allowances: HashMap<Operator, Amount>,
+}
+
+/// The fungible token implementation.
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
+pub struct FTState {
+    total_supply: Amount,
+    owners: HashMap<Owner, OwnerData>,
+}
+
+impl FTState {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Gets the current total token supply.
+    pub const fn total_supply(&self) -> Amount {
+        self.total_supply
+    }
+
+    /// Allows `operator` to spend `amount` of [`msg::source()`]'s tokens.
+    ///
+    /// # Errors
+    /// - [`FTError::ZeroRecipientAddress`] if `operator` is
+    /// [`ActorId::zero()`].
+    pub fn approve(&mut self, operator: Operator, amount: Amount) -> Result<FTApproval, FTError> {
+        self.internal_approve(msg::source(), operator, amount)
+    }
+
+    fn unchecked_internal_approve(&mut self, owner: Owner, operator: Operator, amount: Amount) {
+        self.owners
+            .entry(owner)
+            .or_default()
+            .allowances
+            .insert(operator, amount);
+    }
+
+    fn internal_approve(
+        &mut self,
+        owner: Owner,
+        operator: Operator,
+        amount: Amount,
+    ) -> Result<FTApproval, FTError> {
+        if operator.is_zero() {
+            return Err(FTError::ZeroRecipientAddress);
+        }
+
+        self.unchecked_internal_approve(owner, operator, amount);
+
+        Ok(FTApproval {
+            owner,
+            operator,
+            amount,
+        })
+    }
+
+    /// Returns a balance of `owner`'s tokens.
+    pub fn balance_of(&self, owner: Owner) -> Amount {
```

Comment:
I followed [this standart](https://github.com/w3f/PSPs/blob/master/PSPs/psp-22.md#balance_ofowner-accountid--balance). IMO, the current logic is better, since ERC-20 also doesn't provide information about the `owner` existence.